kafka/consume.go

```diff
@@ -79,39 +85,66 @@ func NewConsumer(log logrus.FieldLogger, conf Config, opts ...ConfigOpt) (*Confl
 		conf.Timeout = DefaultTimout
 	}
 
-	part := int32(-1)
-	if conf.Consumer.Partition != nil {
-		part = *conf.Consumer.Partition
-	}
-
 	cc := &ConfluentConsumer{
-		c:         consumer,
-		conf:      conf,
-		log:       log,
-		partition: part,
+		c:    consumer,
+		conf: conf,
+		log:  log,
 	}
 
-	if cc.partition < 0 {
+	if cc.conf.Consumer.Partition == nil {
 		// if we are not reading from a specific partition, we subscribe
-		cc.configOnce.Do(func() {
-			cc.log.WithField("kafka_topic", cc.conf.Topic).Debug("Subscribing to Kafka topic")
-			cc.rebalanceHandlerMutex.Lock()
-			defer cc.rebalanceHandlerMutex.Unlock()
-			if err := cc.c.Subscribe(cc.conf.Topic, cc.rebalanceHandler); err != nil {
-				err = errors.Wrap(err, "error subscribing to topic")
-			}
-		})
+		cc.setupRebalanceHandler(cc.conf.Consumer.InitialOffset)
+		cc.log.WithField("kafka_topic", cc.conf.Topic).Debug("Subscribing to Kafka topic")
+		cc.rebalanceHandlerMutex.Lock()
+		defer cc.rebalanceHandlerMutex.Unlock()
+		if serr := cc.c.Subscribe(cc.conf.Topic, cc.rebalanceHandler); serr != nil {
+			err = errors.Wrap(serr, "error subscribing to topic")
+		}
 	} else {
 		// if we are reading from a specific partition, we assign
-		cc.configOnce.Do(func() {
-			parts := []kafkalib.TopicPartition{
-				kafkalib.TopicPartition{
-					Topic:     &cc.conf.Topic,
-					Partition: cc.partition,
-				},
+		tps := []kafkalib.TopicPartition{
+			kafkalib.TopicPartition{
+				Topic:     &cc.conf.Topic,
+				Partition: *cc.conf.Consumer.Partition,
+			},
+		}
+		// Set the partition if a key is set to determine the partition
+		if cc.conf.Consumer.PartitionKey != "" && cc.conf.Consumer.PartitionerAlgorithm != "" {
+			cc.SetPartitionByKey(cc.conf.Consumer.PartitionKey, cc.conf.Consumer.PartitionerAlgorithm)
+		}
+		err = cc.c.Assign(tps)
+
+		// handle rebalance events for consumer assigned to a specific partition
+		// - NOTE if the partition is set and the partition key is not set we have no way
+		//   of knowing where to assign the consumer to in the case of a rebalance
+		cc.eventChan = cc.c.Events()
+		go func(cc *ConfluentConsumer) {
```

Comment:
Minor: we should pull this out into a method I think.

Reply:
sure